### ISSUE TITLE
`use_gqa_in_sdpa` doesn't check GPU architecture: up to 28% slower decode on pre-sm80 GPUs

### ISSUE THREAD

**Issue by banarto** [huggingface/transformers#48633]
### System Info

- transformers: 5.0.0
- torch: 2.10.0+cu128
- CUDA: 12.8
- GPU: Tesla T4 (compute capability 7.5)
- Platform: Kaggle Notebook
- Python: 3.12

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

On a T4 (sm75), SDPA always falls back to the MATH backend during decode because `use_gqa_in_sdpa()` returns `True` and `enable_gqa=True` is passed.

https://github.com/huggingface/transformers/blob/361d009f4f6b0659ceccfbeb4888df28374da2a3/src/transformers/integrations/sdpa_attention.py#L27-L36

The function only looks at `head_dim` and whether an attention mask is present — it does not check the GPU architecture. On pre-sm80 devices neither FLASH nor cuDNN kernels exist, so `enable_gqa=True` also removes EFFICIENT_ATTENTION from the candidates, leaving only MATH.

PyTorch reports both facts itself. Running the script below with `set_patch(False)` (stock behaviour) emits:

```
Flash attention only supports gpu architectures in the range [sm80, sm121]. Attempting to run on a sm 7.5 gpu.
```

```
For dense input, both fused kernels require query, key and value to have the same num_heads. Query.sizes(): [1, 12, 1, 128], Key sizes(): [1, 2, 1001, 128]
```

The first says FLASH does not exist here; the second says EFFICIENT is disqualified by the mismatched head counts that `enable_gqa` leaves in place.

### Reproduction script

Runs on a Kaggle T4 notebook. `patched()` repeats K/V before calling `sdpa_attention_forward`, which is equivalent to what an architecture check inside `use_gqa_in_sdpa` would do. Change `name` to `HuggingFaceTB/SmolLM2-360M` to reproduce on a different architecture.

```python
!pip -q install transformers accelerate
import time, torch
from transformers import AutoModelForCausalLM
import transformers.integrations.sdpa_attention as sa
from transformers.modeling_utils import ALL_ATTENTION_FUNCTIONS

# A T4 clocks down when idle. Without this warm-up the same measurement
# varies by 4-6% between runs.
def warmup_gpu(sec=1.0):
    a = torch.randn(4096, 4096, dtype=torch.float16, device="cuda")
    t_end = time.time() + sec
    while time.time() < t_end:
        a @ a
    torch.cuda.synchronize()

def pfl(ids):
    with torch.no_grad():
        return model(ids)

def dcd(prefill_out, n=32):
    past = prefill_out["past_key_values"]
    nxt = torch.argmax(prefill_out["logits"], dim=-1)[:, -1:]
    for _ in range(n):
        with torch.no_grad():
            o = model(nxt, past_key_values=past)
        past = o["past_key_values"]
        nxt = torch.argmax(o["logits"], dim=-1)

# guard against double-patching when the cell is re-run
_cur = sa.sdpa_attention_forward
if getattr(_cur, "_is_patch", False):
    _orig = sa._orig_sdpa_forward
else:
    _orig = _cur
    sa._orig_sdpa_forward = _orig

def patched(module, query, key, value, attention_mask, **kw):
    n_rep = query.shape[1] // key.shape[1]
    if n_rep > 1:
        key = key.repeat_interleave(n_rep, dim=1)
        value = value.repeat_interleave(n_rep, dim=1)
    return _orig(module, query, key, value, attention_mask, **kw)

patched._is_patch = True

def set_patch(on):
    fn = patched if on else _orig
    sa.sdpa_attention_forward = fn
    ALL_ATTENTION_FUNCTIONS["sdpa"] = fn

name = "Qwen/Qwen2.5-1.5B"
model = AutoModelForCausalLM.from_pretrained(
    name, torch_dtype=torch.float16).cuda().eval()

for L in [1000, 4000, 8000]:
    print(f"\n===== ctx={L} =====")
    for on in [False, True]:
        set_patch(on)
        ids = torch.randint(high=1000, size=[1, L], dtype=torch.int, device="cuda")
        out = pfl(ids); dcd(out, n=4)
        ts = []
        for _ in range(5):
            out = pfl(ids)
            warmup_gpu(1.0)
            torch.cuda.synchronize()
            s = torch.cuda.Event(enable_timing=True)
            e = torch.cuda.Event(enable_timing=True)
            s.record(); dcd(out, n=32); e.record()
            torch.cuda.synchronize()
            ts.append(s.elapsed_time(e) / 32)
        label = "repeat_kv" if on else "stock    "
        print(f"  {label} {min(ts):7.2f} ms  ({1000/min(ts):5.1f} tok/s)")

set_patch(False)
```

### Results

Qwen2.5-1.5B (12 Q heads / 2 KV heads)

| ctx  | stock   | with repeat_kv | latency reduction |
|------|---------|----------------|-------------------|
| 1000 | 35.72ms | 29.90ms        | 16.3%             |
| 4000 | 44.90ms | 33.77ms        | 24.8%             |
| 8000 | 71.64ms | 51.26ms        | 28.5%             |

SmolLM2-360M (15 Q heads / 5 KV heads)

| ctx  | stock   | with repeat_kv | latency reduction |
|------|---------|----------------|-------------------|
| 1000 | 38.29ms | 32.44ms        | 15.3%             |
| 4000 | 37.87ms | 31.92ms        | 15.7%             |
| 8000 | 48.78ms | 36.21ms        | 25.8%             |

Each cell is the per-token latency over 32 generated tokens, minimum of five runs. `latency reduction = (stock - patched) / stock`.

Note: the 1000 and 4000 rows are close on a T4 because the GPU is power-limited at these sizes and the extra attention work is absorbed by a clock drop; the difference becomes visible at 8000.

### Correctness

Outputs are unchanged.

- Qwen2.5-1.5B: greedy decoding produces identical token sequences for 200 tokens
- SmolLM2-360M: identical for 50 tokens
- Max absolute logit difference 0.027, mean 0.0034 (fp16)

### Expected behavior

`use_gqa_in_sdpa()` should take the GPU architecture into account and return `False` (i.e. take the `repeat_kv` path) when no fused kernel supporting GQA exists:

```python
if torch.cuda.is_available() and torch.cuda.get_device_capability()[0] < 8:
    return False
```

This check runs on every attention call, so it should probably be cached at module level rather than queried each time.

Not verified: that sm80+ devices are unaffected. Based on kernel availability this should not regress, but I don't have access to such hardware.

Related, both different symptoms:
- #45636 — head_dim > 256 and mask-layout cases; proposes the same `repeat_kv` as part of its fix
- pytorch/pytorch#139586 — notes that `enable_gqa` is only supported by the flash backend

**Comment by banarto** [comment]
Closing as a duplicate. #45776 already covers this, including the `sm < 80` case explicitly, and [this comment](https://github.com/huggingface/transformers/pull/45776#issuecomment-5230620009) has T4 measurements that supersede mine — same device, same torch version, larger effect (1.6–1.8x on batched generate).

That PR also proposes a better fix than the capability check I suggested: using `torch.backends.cuda.can_use_flash_attention()` so the condition tracks FA's actual constraints instead of hardcoding an architecture threshold.

Sorry for the noise — I should have found #45776 before opening this.
